udev: several inotify watch fixups

yuwata opened the pull request:
Hopefully fixes #18525.

Replaces #18524 and #18526.

yuwata commented:
cc @gaoyi1988

bluca commented on src/udev/udev-watch.c:
```diff
@@ -90,12 +91,22 @@ int udev_watch_begin(int inotify_fd, sd_device *dev) {
 
+        r = readlink_malloc(filename, &old_id);
+        if (r >= 0) {
+                if (streq(id, old_id))
```
if I understand correctly these are paths, so path_equal() ?

yuwata replied:
`path_equal()` handles slashes gracefully. But in this case, the device id must be unique, hence should be handled strictly. Moreover, in most cased, the ids saved in symlinks do not look like paths:
```
$ ls -l /run/udev/watch
total 0
lrwxrwxrwx 1 root root 4 Mar  5 08:50 1 -> b8:0
lrwxrwxrwx 1 root root 6 Mar  5 08:49 2 -> b259:2
lrwxrwxrwx 1 root root 6 Mar  5 08:49 3 -> b259:1
lrwxrwxrwx 1 root root 6 Mar  5 08:49 4 -> b259:0
lrwxrwxrwx 1 root root 4 Mar  5 18:41 666 -> b7:0
lrwxrwxrwx 1 root root 4 Mar  5 10:43 70 -> b7:1
```

bluca replied:
I see, makes sense

bluca reviewed (approved):
Changes themselves look good to me, but not that familiar with this area, so waiting for more eyes, and maybe a test run from the reporter

yuwata commented:
Ah, this is completely broken. Will rework...

keszybz commented on src/udev/udev-watch.c:
```diff
@@ -90,12 +91,22 @@ int udev_watch_begin(int inotify_fd, sd_device *dev) {
+                if (streq(id, old_id))
+                        return 0;
+
+                log_device_warning(dev, "'%s' is a symbolic link to a differnt device '%s', ignoring.", filename, old_id);
```
different

yuwata commented:
bionic-i386 fails in portabled test. I hope it is not related. PTAL.

bluca commented:
It doesn't seem unrelated:

```
Mar 08 00:02:30 systemd-testsuite systemd-portabled[406]: Failed to wait for device to be initialized: Connection timed out
Mar 08 00:02:35 systemd-testsuite systemd-portabled[406]: Failed to wait for device to be initialized: Connection timed out
Mar 08 00:02:35 systemd-testsuite systemd-udevd[250]: loop0: Failed to update device symlinks: Too many levels of symbolic links
Mar 08 00:02:36 systemd-testsuite systemd-udevd[247]: loop8: Failed to update device symlinks: Too many levels of symbolic links
```

(as explained in the readme you can download the tarball with the test artifacts, including the journal file, from the linked URL - eg for this case ```wget https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-bionic-upstream-systemd-ci-systemd-ci/bionic/i386/s/systemd-upstream/20210308_002553_0c359@/artifacts.tar.gz```)

gaoyi1988 commented:
@yuwata 
I just backport commit https://github.com/systemd/systemd/pull/18904/commits/70e2dac3362dc88400d63ab6ed403558de41223f and https://github.com/systemd/systemd/pull/18904/commits/647a6321bf85a44611babbeb69879f7fcd365233. And run the test. 

There are some problems:
1. Thers is one device has no symlink path_wd and path_id. so far, it's only happened once.
2. Thers is one device has no symlink path_id. 
```
[root@localhost watch]# ll | grep 49
lrwxrwxrwx. 1 root root 5 Mar  8 17:56 38 -> b8:49
lrwxrwxrwx. 1 root root 5 Mar  8 17:56 49 -> b8:36
lrwxrwxrwx. 1 root root 2 Mar  8 17:56 b8:49 -> 38
```

Are there any other commits I have to backport?

yuwata commented:
Updated. Though, I cannot find the root cause of the previous portable test issue...

@gaoyi1988 Thank you for testing this PR. Could you test again?

bluca commented:
> Updated. Though, I cannot find the root cause of the previous portable test issue...

It will be from the dissect-image logic, ie waiting for the device to appear?

gaoyi1988 commented:
@yuwata 
I have backport https://github.com/systemd/systemd/pull/18904/commits/9c86b95281edd2f3ae05affd6e15797369e2a9b1, https://github.com/systemd/systemd/pull/18904/commits/fb153661705a7bfaf0b6d20ddd0a541e33241ae8 and https://github.com/systemd/systemd/pull/18904/commits/e087790811e0c6a9a8861f188e70cc20b34541f0. But it seems that there are possibilities that some devices have no symlink path_id.
```
[root@localhost watch]# ll | grep 47
lrwxrwxrwx. 1 root root 5 Mar  9 20:58 47 -> b8:65
```

gaoyi1988 commented:
@yuwata 
Accordding to the logs, no path_id is happened because of change event and inotify event with IN_IGNORED are both happened, and symlink path_id is deleted by inotify event which created by change event.
```
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: Processing device (SEQNUM=398878, ACTION=change)
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: Removing watch
Mar 10 11:32:00 localhost.localdomain **systemd-udevd[79288]: dm-2: Inotify event: 8000 for /dev/dm-2**
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: Delete watch handle 12, b253:2
...
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: Adding watch on '/dev/dm-2'
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: Delete watch handle -2, b253:2
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: **dm-2: Create watch handle /run/udev/watch/65 /run/udev/watch/b253:2**
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: sd-device: Created db file '/run/udev/data/b253:2' for '/devices/virtual/block/dm-2'
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: Device (SEQNUM=398878, ACTION=change) processed
Mar 10 11:32:00 localhost.localdomain systemd-udevd[79335]: dm-2: sd-device-monitor: Passed 889 byte to netlink monitor
Mar 10 11:32:00 localhost.localdomain **systemd-udevd[79288]: dm-2: Delete watch handle -2, b253:2**
```

yuwata commented:
@gaoyi1988 Updated. Could you test again?

yuwata commented:
> dm-2: Delete watch handle -2, b253:2

Where does this come from?

gaoyi1988 commented:
> > dm-2: Delete watch handle -2, b253:2
> 
> Where does this come from?

"Create watch handle..." and "Delete watch handle..." is added by me for debug. 
In the end of device_remove_watch_handle, add:
```
log_device_debug(device, "Delete watch handle %d, %s", wd, id); 
```
In device_set_watch_handle before call symlink, add:
```
log_device_debug(device, "Create watch handle %s %s", path_wd, path_id);
```
In that logs, PID 79288 is main process and handles inotify event, PID 79335 handles change event.

gaoyi1988 commented:
@yuwata 
I think the real problem is in function device_remove_watch_handle. If we delete path_wd, we must delete path_id at the same time. But if we can't find path_wd, path_id may be updated by other event, so we can't delete.

I think your commit cf457241e3bba83f5804f0f38d415587c593168c didn't fix this problem.

yuwata commented:
Updated.

@gaoyi1988 
> In that logs, PID 79288 is main process and handles inotify event, PID 79335 handles change event.

Yeah, calling `udev_watch_end()` from the main process may cause race condition. I dropped it. Now, `udev_watch_begin/end()` are only called from udev worker processs. Hence, there should be no race anymore. Please test again. Thank you for your help!

yuwata commented:
@gaoyi1988 BTW, how do you test this PR? I included your original reproducer reported in #18525 in our test. If you find any issues in the current PR, then please provide the reproducer.

yuwata commented:
Wow! Finally, all CIs are green!

gaoyi1988 commented:
@yuwata 
Delete udev_watch_end in on_inotify can fix the problem. And my testcase is in https://github.com/systemd/systemd/issues/18525

yuwata commented:
Thank you!

gaoyi1988 commented:
@yuwata 
Are there still any problems? Why this is not merged?

yuwata commented:
Because no one has reviewed this yet, and we are slightly busy for v248 release process. I hope this will be merged soon after v248 is released.

poettering commented on src/libsystemd/sd-device/sd-device.c:
```diff
@@ -775,60 +792,45 @@ static int device_set_drivers_subsystem(sd_device *device, const char *_subsyste
+                        return log_device_debug_errno(device, r,
+                                                      "sd-device: Failed to read subsystem for %s: %m",
+                                                      device->devpath);
                 if (r >= 0)
```
maybe change this to `if (subsystem)` to make this more readable? doesn't matter though

poettering commented on src/libsystemd/sd-device/device-private.h:
```diff
@@ -12,18 +12,22 @@
 void device_seal(sd_device *device);
 void device_set_is_initialized(sd_device *device);
-void device_set_watch_handle(sd_device *device, int fd);
+int device_set_watch_handle(sd_device *device, int handle);
```
we usually call these "wd"

poettering commented on src/libsystemd/sd-device/sd-device.c:
```diff
@@ -1239,10 +1239,7 @@ static int handle_db_line(sd_device *device, char key, const char *value) {
-                if (r < 0)
-                        return r;
-
+                /* Deprecated. */
```
what happens during upgrade from old udev?

yuwata replied:
Watch handle is always read from /run/udev/watch now. New udevd simply ignores the handle saved in udev databse.

poettering commented on src/udev/udevd.c:
```diff
@@ -1293,17 +1293,18 @@ static int on_inotify(sd_event_source *s, int fd, uint32_t revents, void *userda
                 log_device_debug(dev, "Inotify event: %x for %s", e->mask, devnode);
                 if (e->mask & IN_CLOSE_WRITE)
-                        synthesize_change(dev);
-                else if (e->mask & IN_IGNORED)
```
hmm, why is the IN_IGNORED handling dropped?

gaoyi1988 replied:
@yuwata 
This is also my question. 

Actually, before verified this solution, I have verified that delete path_id and path_wd both, or delete anything can solve the problem. And I think this is the real problem. As follow:
```
static void device_remove_watch_handle(sd_device *device) {
        const char *id;
        int wd;

        assert(device);

        /* First, remove the symlink from handle to device id. */
        wd = device_get_watch_handle(device);
        if (wd >= 0) {
                char path_wd[STRLEN("/run/udev/watch/") + DECIMAL_STR_MAX(int)];

                xsprintf(path_wd, "/run/udev/watch/%d", wd);
                if (access(path_wd, F_OK) < 0)
                        return;
                
                (void) unlink(path_wd);
        } else 
                return;

        /* Next, remove the symlink from device id to handle. */
        if (device_get_id_filename(device, &id) >= 0) {
                const char *path_id;

                path_id = strjoina("/run/udev/watch/", id);
                (void) unlink(path_id);
        }

        device->watch_handle = -1;
}
```

yuwata replied:
When `IN_IGNORED` is triggered, the corresponding 'remove' uevent for the device will be triggered, and during the process of the uevent, `udev_watch_end()` is called. So, it is not necessary to handle the flag here. Moreover, when both `udev_watch_end()` is called by both the main and worker process, then `device_remove_watch_handle()` in `udev_watch_end()` may cause some race. With this commit, `udev_watch_begin()` and `udev_watch_end()` are only called by workers. So, there should be no race.

I added short comment about that in the code.

poettering reviewed (commented):
looks good, a few comments

yuwata commented on src/udev/udev-event.c:
```diff
@@ -993,10 +995,8 @@ int udev_event_execute_rules(
 
-        if (action == SD_DEVICE_REMOVE) {
-                event_execute_rules_on_remove(event, inotify_fd, timeout_usec, timeout_signal, properties_list, rules);
-                return 0;
```
When, some remove rules are failed, then this commit changes only two things: warn the failure, and skip `udev_event_execute_run()`. Skipping `udev_event_execute_run()` on some failure should be safer, as the execution list may be different from expected.

yuwata commented:
@poettering Thank you for your review. Updated. PTAL.

poettering commented on src/libsystemd/sd-device/sd-device.c:
```diff
@@ -751,15 +751,32 @@ int device_set_subsystem(sd_device *device, const char *_subsystem) {
+        if (p <= syspath)
+                return -EINVAL;
+        p++;
+        if (p >= drivers)
```
maybe add comment here: `/* refuse duplicate slashes */` or so? because that's what this is supposed to do, no?

poettering commented on src/libsystemd/sd-device/device-private.c:
```diff
@@ -568,28 +569,132 @@ int device_get_devlink_priority(sd_device *device, int *priority) {
+                const char *path_id;
+
+                path_id = strjoina("/run/udev/watch/", id);
+                (void) unlink(path_id);
```
doesn't matter, but maybe debug log if this fails with errors != ENOENT? 

doens't matter at all though

poettering commented on src/libsystemd/sd-device/device-private.c:
```diff
@@ -568,28 +569,132 @@ int device_get_devlink_priority(sd_device *device, int *priority) {
+        assert(ret);
+
+        if (wd < 0)
+                return -EINVAL;
```
EBADF? (i mean, that's the error you return further up when the wd db on disk is borked)

poettering reviewed (approved):
looks all great to me. only found nitpicks

yuwata commented:
@poettering Thank you for review. All the above comments are addressed. Upgrading the green label.
